# Comparative between DFT and FFt

In [1]:
import numpy as np
from scipy.fft import fft, fftfreq
import matplotlib.pyplot as plt

In [2]:
def DFT_slow(x):
    """Compute the discrete Fourier Transform of the 1D array x"""
    x = np.asarray(x, dtype=float)
    N = x.shape[0]
    n = np.arange(N)
    k = n.reshape((N, 1))
    M = np.exp(-2j * np.pi * k * n / N)
    return np.dot(M, x)

In [3]:
x = np.random.random(1024)
np.allclose(DFT_slow(x), np.fft.fft(x))

True

In [15]:
%timeit DFT_slow(x)
%timeit np.fft.fft(x)

10 loops, best of 5: 125 ms per loop
The slowest run took 43.56 times longer than the fastest. This could mean that an intermediate result is being cached.
10000 loops, best of 5: 17.7 µs per loop


We are over 1000 times slower, which is to be expected for such a simplistic implementation. But that's not the worst of it. For an input vector of length N, the FFT algorithm scales as O[NlogN], while our slow algorithm scales as O[N2]. That means that for N=106 elements, we'd expect the FFT to complete in somewhere around 50 ms, while our slow algorithm would take nearly 20 hours!

https://jakevdp.github.io/blog/2013/08/28/understanding-the-fft/

In [11]:
import time
inicio = time.time()

# Código a medir
DFT_slow(x)

fin = time.time()
print(fin-inicio)

0.0935366153717041


In [12]:
import time
inicio = time.time()

# Código a medir
np.fft.fft(x)

fin = time.time()
print(fin-inicio)

0.000278472900390625


In [16]:
a = 0.0935366153717041/0.000278472900390625
a

335.89126712328766

In this example we can see that the **DFT** is over 335.89126712328766 slower than **numpy fft** function what is more efficient because this one takes advantage of the symetry of the signal.